pkg/services/apiserver/builder/metrics.go

```
@@ -0,0 +1,34 @@
+package builder
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type builderMetrics struct {
+	dualWriterMode *prometheus.CounterVec
+}
+
+func newBuilderMetrics(reg prometheus.Registerer) *builderMetrics {
+	return &builderMetrics{
+		dualWriterMode: promauto.With(reg).NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "dual_writer_mode",
+				Help: "Dual writer mode",
+			},
+			[]string{"resource", "group", "targetMode", "currentMode"},
```

Could the current mode be *value* of the metric, instead of label? What is "target mode"?

---

So, basically we are trying to track if the dual writer was able to execute its operations and thus was able to update the mode of the dual writer successfully. We have encountered cases where it failed silently and caused issues.

As you mentioned this I thought a bit more on this topic. It seems to me that 2 metrics would make more sense perhaps in this case. Let me broaden this and let me know what do you think?

`dual_writer_target_mode` and `dual_writer_current_mode` (both could be `GaugeVec`) where we expect them to be same value if the process has completed successfully.

---

That makes sense to me. For my own learning, I wonder why target and current mode would be different?

---

Thanks for providing context and links!